File: frontend/services/agent.service.ts
import { Complaint } from "@/types";
import { apiClient } from "./api.client";

/**
 * Get agent's complaints (filtered by municipality)
 */
export const getAgentComplaints = async (params?: {
  status?: string;
  category?: string;
  page?: number;
  limit?: number;
}): Promise<{
  success: boolean;
  message: string;
  data: {
    complaints: Complaint[];
    pagination: {
      total: number;
      page: number;
      limit: number;
      pages: number;
    };
    municipalityName?: string;
  };
}> => {
  const searchParams = new URLSearchParams();
  if (params?.status) searchParams.set("status", params.status);
  if (params?.category) searchParams.set("category", params.category);
  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.limit) searchParams.set("limit", params.limit.toString());

  const queryString = searchParams.toString();
  const endpoint = `/agent/complaints${queryString ? `?${queryString}` : ""}`;

  return apiClient.get<{
    success: boolean;
    message: string;
    data: {
      complaints: Complaint[];
      pagination: {
        total: number;
        page: number;
        limit: number;
        pages: number;
      };
    };
  }>(endpoint);
};

/**
 * Validate a submitted complaint
 */
export const validateComplaint = async (
  id: string
): Promise<{ success: boolean; message: string; data: Complaint }> => {
  return apiClient.put<{ success: boolean; message: string; data: Complaint }>(
    `/agent/complaints/${id}/validate`,
    {}
  );
};

/**
 * Reject a submitted complaint with justification
 */
export const rejectComplaint = async (
  id: string,
  reason: string
): Promise<{ success: boolean; message: string; data: Complaint }> => {
  return apiClient.put<{ success: boolean; message: string; data: Complaint }>(
    `/agent/complaints/${id}/reject`,
    { reason }
  );
};

/**
 * Close a resolved complaint
 */
export const closeComplaint = async (
  id: string
): Promise<{ success: boolean; message: string; data: Complaint }> => {
  return apiClient.put<{ success: boolean; message: string; data: Complaint }>(
    `/agent/complaints/${id}/close`,
    {}
  );
};

/**
 * Get available departments
 */
export const getAgentDepartments = async (): Promise<{
  success: boolean;
  data: Array<{ _id: string; name: string; email?: string; phone?: string }>;
}> => {
  return apiClient.get<{
    success: boolean;
    data: Array<{ _id: string; name: string; email?: string; phone?: string }>;
  }>("/agent/departments");
};

/**
 * Assign complaint to department
 */
export const assignComplaintToDepartment = async (
  complaintId: string,
  departmentId: string
): Promise<{ success: boolean; message: string; data: Complaint }> => {
  return apiClient.put<{ success: boolean; message: string; data: Complaint }>(
    `/agent/complaints/${complaintId}/assign-department`,
    { departmentId }
  );
};

/**
 * Get agent statistics
 */
export const getAgentStats = async (): Promise<{
  success: boolean;
  data: {
    total: number;
    submitted: number;
    validated: number;
    assigned: number;
    inProgress: number;
    resolved: number;
    closed: number;
    rejected: number;
    totalOverdue: number;
    totalAtRisk: number;
    resolutionRate: number;
    averageResolutionTime: number;
    byCategory: Record<string, number>;
    byMonth: Record<string, number>;
  };
}> => {
  return apiClient.get("/complaints/stats");
};

/**
 * Approve technician resolution report
 */
export const approveResolution = async (
  complaintId: string
): Promise<{ success: boolean; message: string; data: Complaint }> => {
  return apiClient.post<{ success: boolean; message: string; data: Complaint }>(
    `/agent/complaints/${complaintId}/approve-resolution`,
    {}
  );
};

/**
 * Reject technician resolution report
 */
export const rejectResolution = async (
  complaintId: string,
  rejectionReason: string
): Promise<{ success: boolean; message: string; data: Complaint }> => {
  return apiClient.post<{ success: boolean; message: string; data: Complaint }>(
    `/agent/complaints/${complaintId}/reject-resolution`,
    { rejectionReason }
  );
};

export const agentService = {
  getAgentComplaints,
  validateComplaint,
  rejectComplaint,
  closeComplaint,
  getAgentDepartments,
  assignComplaintToDepartment,
  getStats: getAgentStats,
  approveResolution,
  rejectResolution,
};

// AI Department Prediction
export const predictDepartment = async (category: string, description: string, municipality?: string): Promise<{
  success: boolean;
  data?: {
    suggestedDepartment: string;
    departmentName: string;
    confidence: number;
    message: string;
  };
}> => {
  return apiClient.post("/public/ai/predict-department", { category, description, municipality });
};

File: mobile/lib/services/agent_service.dart
import 'api_client.dart';

class AgentService {
  final ApiClient _apiClient = ApiClient();

  Future<List<dynamic>> getAgentComplaints({
    String? status,
    int page = 1,
    int limit = 50,
  }) async {
    String endpoint = '/agent/complaints?page=$page&limit=$limit';
    if (status != null && status.isNotEmpty && status != 'ALL') {
      endpoint += '&status=$status';
    }
    final response = await _apiClient.get(endpoint);
    final data = response?['data'];
    if (data is Map) return data['complaints'] ?? [];
    if (data is List) return data;
    return response?['complaints'] ?? [];
  }

  Future<void> validateComplaint(String id) async {
    await _apiClient.put('/agent/complaints/$id/validate', {});
  }

  Future<void> rejectComplaint(String id, String reason) async {
    await _apiClient.put('/agent/complaints/$id/reject', {'reason': reason});
  }

  Future<void> assignDepartment(String id, String departmentId) async {
    await _apiClient.put('/agent/complaints/$id/assign-department', {
      'departmentId': departmentId,
    });
  }

  Future<Map<String, dynamic>> predictDepartment(
    String category,
    String description,
    String municipality,
  ) async {
    final response = await _apiClient.post('/ai/predict-department', {
      'category': category,
      'description': description,
      'municipality': municipality,
    });
    return response ?? {};
  }

  Future<void> approveResolution(String id, {String? notes}) async {
    await _apiClient.post('/agent/complaints/$id/approve-resolution', {
      if (notes != null && notes.trim().isNotEmpty) 'notes': notes,
    });
  }

  Future<void> rejectResolution(String id, {String? reason}) async {
    await _apiClient.post('/agent/complaints/$id/reject-resolution', {
      if (reason != null && reason.trim().isNotEmpty) 'reason': reason,
    });
  }

  Future<List<dynamic>> getDepartments() async {
    final response = await _apiClient.get('/agent/departments');
    return response?['data'] ?? response?['departments'] ?? [];
  }

  Future<Map<String, dynamic>> getStats() async {
    final response = await _apiClient.get('/agent/stats');
    return response?['data'] ?? {};
  }
}
